### ISSUE TITLE
Tpetra::DistObject: DualView arguments have wrong types

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#4344]
@trilinos/tpetra @kyungjoo-kim 

This concerns the following DistObject virtual methods that its subclasses must implement:

  - copyAndPermuteNew
  - packAndPrepareNew
  - unpackAndCombineNew

As of PR #4328, we have clarified how `DistObject::doTransferNew` expects these methods to behave, with respect to their DualView parameters.  Export and Import setup is responsible for ensuring that `permuteToLIDs`, `permuteFromLIDs`, `exportLIDs`, and `remoteLIDs` are sync'd on both host and device.  Thus, the above methods need not sync any of these four DualViews.  As a result, their type `const DualView<const LO*, buffer_device_type>&` is correct.  (`const DualView&` means you can't sync them; `DualView<const LO*, ...>` means you can't modify their entries.)

PR #4328 further clarifies that `packAndPrepareNew` may fill its `exports` and `numPacketsPerLID` DualView parameters wherever the subclass likes, either on host or device.  It is not responsible for sync'ing either of these two DualViews to any particular place; `DistObject::doTransferNew` must sync them wherever it needs them to be.  Similarly, `unpackAndCombineNew` may unpack received data (`imports` and `numPacketsPerLIDs`) wherever it wants, either on host or device, but is responsible for sync'ing them wherever it needs them.  `DistObject::doTransferNew` is not responsible for sync'ing them.

I justify the previous paragraph as follows:

  - Avoid unnecessary syncs to host, if the subclass knows how to unpack on device
  - Separation of concerns between communication and pack / unpack

This imposes the following requirements on `exports`, `imports`, and `numPacketsPerLIDs`:

  1. `DistObject::doTransferNew` allocates the `numPacketsPerLIDs` argument when it calls `packAndPrepareNew`, but `packAndPrepareNew` must allocate the `exports` output argument.  Thus, subclasses cannot reallocate `numPacketsPerLID`.
  2. `DistObject::doTransferNew` allocates the `imports` and `numPacketsPerLIDs` arguments when it calls `unpackAndCombineNew`; subclasses cannot reallocate them.
  3. Subclasses need to be able to sync any of these arrays wherever they like.
  4. `packAndPrepareNew` implementations access `exports` and `numPacketsPerLIDs` in write-only fashion.
  5. `unpackAndPrepareNew` implementations technically only need read-only access to `imports` and `numPacketsPerLIDs`.  However, `DistObject::doTransferNew` only accesses them in write-only fashion before calling `unpackAndPrepareNew`, and doesn't access them afterwards.  Thus, `unpackAndPrepareNew` implementations could freely write to these arrays, e.g., to use them as scratch space.

This means that the correct type of `imports` and `numPacketsPerLIDs` is `DualView<T*, ...>`.

  - Can't have `DualView<const T*, ...>`, because that would forbid sync'ing.
  - Can't have `const DualView&` or `const DualView`, because that would forbid sync'ing.
  - Can't have `DualView&`, because that would expose DistObject's internal DualViews to reallocation by subclasses.  Passing by value means that even if the subclass reallocates, the caller won't see that.  It's just like passing `double*` into a function, instead of `double*&`.

Currently, these arrays have type `const DualView<const T*, ...>&`.  This makes implementations do hacks in order to meet the above four requirements (esp. (2)).

I propose changing the types of `imports`, and `numPacketsPerLID` (as parameters to `packAndPrepareNew` and `unpackAndCombineNew`) from `const DualView<const T*, ...>&`, to `DualView<T*, ...>`.

**Comment by mhoemmen** [comment]
I have a fix in progress, but it depends on PR #4328.  I'd rather get that one in first, since it should improve performance.

**Comment by jhux2** [comment]
@william76 Did you intend to close this?

**Comment by mhoemmen** [comment]
@jhux2 It's good to close; the fix is in master.  I think closing issues once their fixing commit merges into master is automatic.  It shows up as @william76 because he does the merges.

**Comment by william76** [comment]
@jhux2 @mhoemmen
Yeah, this caught me off guard a while back too. When I merge the develop branch into master for some reason github closes these open & merged PR's and says I did it... /shrug 
It's probably good anyhow for housekeeping... if the information in the PR is needed to live past the time it's been merged the information should be put into an issue.

We do need to try and keep the # of open PR's relatively low since the amount of communication to Github from the autotester scales with the # of PR's that are open and Github has a cap on that which, if hit, will make github ignore it until the counter resets... it's happened a couple times and it caused a bunch of PR's to fail.  That was why I went through before the winter shutdown and closed out a bunch of old PR's that had been 'open' for over a year ;)
